core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -159,48 +159,63 @@ sealed trait PartitionState {
 }
 
 sealed trait PendingPartitionChange extends PartitionState {
+  def lastCommittedState: PartitionState
```

Comment:
> Is the type PartitionState too flexible? Should this always be CommittedPartitionState?

I think that it makes sense to use `CommittedPartitionState` here. I don't think that it is necessary to move `lastCommittedState` to `CommittedPartitionState` though because we can only revert based on a `PendingPartitionChange`.

> Now that we have lastCommittedState, I think that we have redundant data in both PendingPartitionChange types. For example, isr is always set to lastCommittedState.isr.

Yeah, that is right. Let me consolidate. For now, I have changed `newLeaderAndIsr`. I will take a look at this as followup.